File: immi-pulse-be/src/app/agents/immigration/questionnaires/rules_engine.py
"""Conditional-logic rules engine for questionnaire fields.

Pure functions — no DB, no I/O. Mirror this contract exactly in the
frontend TypeScript port (src/lib/questionnaires/rulesEngine.ts) so client
and server agree on which fields are visible.

A field may carry `logic` (visibility + required-if rules). At fill time
and at submission time, we walk the field list in order and compute which
fields are visible. Fields hidden by their rules:
  - never count toward "required" validation
  - have their answers dropped before persistence

A Rule references another field's key. Cycles are impossible because rules
can only reference fields that appear *earlier* in the form (enforced at
authoring time by the UI; defensively allowed here — we treat unknown or
not-yet-evaluated keys as having no answer).
"""

from __future__ import annotations

from typing import Any, Iterable


OPERATORS = (
    "equals",
    "not_equals",
    "is_one_of",
    "is_not_one_of",
    "is_empty",
    "is_not_empty",
    "contains",
)

VISIBILITY_MODES = ("always", "show_if", "hide_if")


def _is_blank(v: Any) -> bool:
    if v is None:
        return True
    if isinstance(v, str) and v.strip() == "":
        return True
    if isinstance(v, (list, tuple, dict)) and len(v) == 0:
        return True
    return False


def _as_list(v: Any) -> list:
    if v is None:
        return []
    if isinstance(v, list):
        return v
    return [v]


def evaluate_rule(rule: dict, answers: dict[str, Any]) -> bool:
    """Evaluate a single rule against the current answers dict.

    Rule shape: {"field_key": str, "operator": str, "value": Any}.
    Returns False for unknown operators (fail-safe — don't accidentally
    show urgent fields from a malformed rule).
    """
    key = rule.get("field_key")
    op = rule.get("operator")
    expected = rule.get("value")
    if not key or op not in OPERATORS:
        return False
    actual = answers.get(key)

    if op == "is_empty":
        return _is_blank(actual)
    if op == "is_not_empty":
        return not _is_blank(actual)
    if op == "equals":
        return actual == expected
    if op == "not_equals":
        return actual != expected
    if op == "is_one_of":
        return actual in _as_list(expected)
    if op == "is_not_one_of":
        return actual not in _as_list(expected)
    if op == "contains":
        if isinstance(actual, str) and isinstance(expected, str):
            return expected.lower() in actual.lower()
        if isinstance(actual, list):
            return expected in actual
        return False
    return False


def evaluate_rules_all(rules: Iterable[dict], answers: dict[str, Any]) -> bool:
    """AND-composition: True only when every rule passes (or no rules)."""
    rules = list(rules)
    if not rules:
        return True
    return all(evaluate_rule(r, answers) for r in rules)


def is_field_visible(field: dict, answers: dict[str, Any]) -> bool:
    """Return True if `field` should be shown given `answers`.

    Visibility config:
      logic.visibility = {"mode": "always" | "show_if" | "hide_if",
                          "rules": [Rule]}
    Default (no logic block) is "always visible".
    """
    logic = field.get("logic") or {}
    visibility = logic.get("visibility") or {}
    mode = visibility.get("mode") or "always"
    rules = visibility.get("rules") or []

    if mode == "always":
        return True
    if mode == "show_if":
        return evaluate_rules_all(rules, answers)
    if mode == "hide_if":
        return not evaluate_rules_all(rules, answers)
    return True


def is_field_required(field: dict, answers: dict[str, Any]) -> bool:
    """Return True if the field is required given the current answers.

    Required = base `required` flag OR every `required_if` rule passes.
    A hidden field is never required (caller should check visibility first
    but we also short-circuit here for safety).
    """
    if not is_field_visible(field, answers):
        return False
    if field.get("required"):
        return True
    logic = field.get("logic") or {}
    required_if = logic.get("required_if") or []
    if not required_if:
        return False
    return evaluate_rules_all(required_if, answers)


def visible_field_keys(fields: list[dict], answers: dict[str, Any]) -> set[str]:
    """Compute the set of visible field keys for the given answers."""
    return {f["key"] for f in fields if is_field_visible(f, answers)}


def clean_answers(
    fields: list[dict], answers: dict[str, Any]
) -> dict[str, Any]:
    """Strip answers for fields that are not visible. Defends against
    tampered or stale submissions where the client sent values for hidden
    fields. Unknown answer keys (no matching field) are also dropped.
    """
    visible = visible_field_keys(fields, answers)
    return {k: v for k, v in answers.items() if k in visible}


def missing_required(
    fields: list[dict], answers: dict[str, Any]
) -> list[str]:
    """Return labels of required-and-visible fields that have no answer."""
    out: list[str] = []
    for f in fields:
        if not is_field_required(f, answers):
            continue
        if _is_blank(answers.get(f["key"])):
            out.append(f.get("label") or f.get("key") or "Required field")
    return out

File: immi-pulse-fe/src/lib/questionnaires/rulesEngine.ts
/**
 * Conditional-logic rules engine for questionnaire fields — TypeScript port.
 *
 * MUST mirror the Python implementation in
 *   immi-pulse-be/src/app/agents/immigration/questionnaires/rules_engine.py
 * exactly. Both sides share a fixture
 *   immi-pulse-be/tests/fixtures/rule_eval_cases.json
 * so divergence is caught in either pytest or vitest.
 */

import type {
  FieldLogic,
  QuestionField,
  Rule,
  RuleOperator,
} from "@/lib/api/services";

const OPERATORS: RuleOperator[] = [
  "equals",
  "not_equals",
  "is_one_of",
  "is_not_one_of",
  "is_empty",
  "is_not_empty",
  "contains",
];

export type Answers = Record<string, unknown>;

const isBlank = (v: unknown): boolean => {
  if (v === undefined || v === null) return true;
  if (typeof v === "string" && v.trim() === "") return true;
  if (Array.isArray(v) && v.length === 0) return true;
  if (typeof v === "object" && v !== null && Object.keys(v).length === 0)
    return true;
  return false;
};

const asList = (v: unknown): unknown[] => {
  if (v === undefined || v === null) return [];
  if (Array.isArray(v)) return v;
  return [v];
};

export function evaluateRule(rule: Rule, answers: Answers): boolean {
  const { field_key, operator, value: expected } = rule;
  if (!field_key || !OPERATORS.includes(operator)) return false;
  const actual = answers[field_key];

  switch (operator) {
    case "is_empty":
      return isBlank(actual);
    case "is_not_empty":
      return !isBlank(actual);
    case "equals":
      return actual === expected;
    case "not_equals":
      return actual !== expected;
    case "is_one_of":
      return asList(expected).includes(actual as never);
    case "is_not_one_of":
      return !asList(expected).includes(actual as never);
    case "contains":
      if (typeof actual === "string" && typeof expected === "string") {
        return actual.toLowerCase().includes(expected.toLowerCase());
      }
      if (Array.isArray(actual)) {
        return actual.includes(expected as never);
      }
      return false;
    default:
      return false;
  }
}

export function evaluateRulesAll(
  rules: Rule[] | undefined,
  answers: Answers,
): boolean {
  if (!rules || rules.length === 0) return true;
  return rules.every((r) => evaluateRule(r, answers));
}

export function isFieldVisible(
  field: QuestionField,
  answers: Answers,
): boolean {
  const logic: FieldLogic | null | undefined = field.logic;
  const visibility = logic?.visibility;
  const mode = visibility?.mode ?? "always";
  const rules = visibility?.rules ?? [];

  if (mode === "always") return true;
  if (mode === "show_if") return evaluateRulesAll(rules, answers);
  if (mode === "hide_if") return !evaluateRulesAll(rules, answers);
  return true;
}

export function isFieldRequired(
  field: QuestionField,
  answers: Answers,
): boolean {
  if (!isFieldVisible(field, answers)) return false;
  if (field.required) return true;
  const requiredIf = field.logic?.required_if ?? [];
  if (requiredIf.length === 0) return false;
  return evaluateRulesAll(requiredIf, answers);
}

export function visibleFieldKeys(
  fields: QuestionField[],
  answers: Answers,
): Set<string> {
  const out = new Set<string>();
  for (const f of fields) {
    if (isFieldVisible(f, answers)) out.add(f.key);
  }
  return out;
}

export function cleanAnswers(
  fields: QuestionField[],
  answers: Answers,
): Answers {
  const visible = visibleFieldKeys(fields, answers);
  const result: Answers = {};
  for (const k of Object.keys(answers)) {
    if (visible.has(k)) result[k] = answers[k];
  }
  return result;
}

export function missingRequired(
  fields: QuestionField[],
  answers: Answers,
): string[] {
  const out: string[] = [];
  for (const f of fields) {
    if (!isFieldRequired(f, answers)) continue;
    if (isBlank(answers[f.key])) {
      out.push(f.label || f.key || "Required field");
    }
  }
  return out;
}

/* --- Authoring-time validators (used by the RuleEditor UI) --- */

/**
 * Returns a list of error messages explaining why a set of fields wouldn't
 * pass server-side validation. Empty array = valid. Used to render inline
 * warnings under a misconfigured field.
 */
export function validateFieldRules(
  fields: QuestionField[],
): { fieldKey: string; message: string }[] {
  const errors: { fieldKey: string; message: string }[] = [];
  const seenKeys = new Set<string>();
  const dupes = new Set<string>();
  for (const f of fields) {
    if (seenKeys.has(f.key)) dupes.add(f.key);
    seenKeys.add(f.key);
  }
  for (const dup of dupes) {
    errors.push({
      fieldKey: dup,
      message: `Duplicate field key "${dup}". Each field needs a unique key.`,
    });
  }
  const seenSoFar = new Set<string>();
  for (const f of fields) {
    const rules = [
      ...(f.logic?.visibility?.rules ?? []),
      ...(f.logic?.required_if ?? []),
    ];
    for (const r of rules) {
      if (r.field_key === f.key) {
        errors.push({
          fieldKey: f.key,
          message: `Rule references the field itself.`,
        });
      } else if (!seenKeys.has(r.field_key)) {
        errors.push({
          fieldKey: f.key,
          message: `Rule references unknown field "${r.field_key}".`,
        });
      } else if (!seenSoFar.has(r.field_key)) {
        errors.push({
          fieldKey: f.key,
          message: `Rule references "${r.field_key}", which appears later in the form. Move it above this field.`,
        });
      }
    }
    seenSoFar.add(f.key);
  }
  return errors;
}
